README.md
# Virtual Directory for Shelf

![Pub Version](https://img.shields.io/pub/v/shelf_virtual_directory)

This package provides `Handler`, `Router` and `Cascade` to serve static files from the [*dart shelf server*](https://pub.dev/packages/shelf) and can work with websockets.

| Provider  | UseCase                                                           |
| --------- | ----------------------------------------------------------------- |
| `Handler` | Can be used to serve as a handler for `Pipeline` or for `Cascade` |
| `Router`  | Can be used to mount as a subroute                                |
| `Cascade` | Can be used directly serve from shelf server                      |

## Setup

1) Add to pubspec.yaml file

    ```yaml
    dependencies:
        shelf_virtual_directory: latest_version
    ```

2) Get dependencies

    ```shell
    pub get
    ```

3) Import

   ```dart
   import 'package:shelf_virtual_directory/shelf_virtual_directory.dart';
   import 'package:path/path.dart' as p;
   ```

4) Create a instance of `ShelfVirtualDirectory` with a directory `web`

   ```dart
   final fsPath = p.join(Directory.current.path,'example','web');// path to server
   final virtualDir = ShelfVirtualDirectory(
        fsPath,
        defaultFile:'index.html',
        default404File:'404.html',
    );
   ```

   **Note: `index.html` and `404.html` files must be directly under root folder. In above case under `web/` folder.**

## Handling different cases

- Using as a `Handler`

    ```dart
    final virDirHandler = ShelfVirtualDirectory(fsPath).handler;

    final staticFileHandler = const Pipeline()
        .addMiddleware(logRequests())
        .addHandler(virDirHandler);//used as a handler

    io.serve(Cascade().add(staticFileHandler).handler,address,port)
    .then((server){
        print('Server is sunning at ${server.address}:${server.port}'),
    });
    ```

- Using as a `Router`

    ```dart
    //As a subroute
    final router = Router()
    ..get('/otherroute',otherRoutehandler)
    ..mount('/',ShelfVirtualDirectory(fsPath).router); // localhost:8080/
    //or

    final router = Router()
    ..get('/otherroute',otherRouteHandler)
    ..mount('/home/',ShelfVirtualDirectory(fsPath).router);//localhost:8080/home/
    ```

    **Note: If your are planning to use it under home directory('/'), always mount or handle the `ShelfVirtualDirectory` at the end.**

    Example

    ```dart
    final mainRoute = Router()
    ..get('/rest',(_)=>Respond.ok('Other routes'))
    ..mount('/',ShelfVirtualDirectory(fsPath).router);
    ```

- Using as a `Cascade`

    ```dart
    import 'package:shelf/shelf_io.dart' as io show serve;

    // You can add other handlers to this cascade
    final virDirCascade = ShelfVirtualDirectory(fsPath).cascade;
    io.serve(virDirCascade.add(someOtherHandler),'localhost',8080)
    .then((server){
      print('Server is sunning at ${server.address}:${server.port}'),
    })
    ```

## Limitations

- Only `mount` method supports for the router.

    ```dart
    final router = Router()
        ..mount('/home/',virDir.handler)
        ..get('/api',apiHandler);
    ```

## Contrubitions

**All contrubitions are welcomed.**


CHANGELOG.md
# Change Log

## 3.0.1

- Upgraded dependencies
- Fixed linter warnings

## 3.0.0

Complete rewrite of this package.

### Major changes

- Removed [`showLogs`](https://github.com/RatakondalaArun/shelf_virtual_directory/blob/v2.0.0/lib/src/virtual_directory.dart#L78) argument.
- Now `folderPath` takes a [absolute path](https://www.hackterms.com/absolute%20path).

### Minor changes

- Upgrade dependencies
- Added support for range requests [#19](https://github.com/RatakondalaArun/shelf_virtual_directory/pull/19).
- Issues Fixed
  - New files does not get served untill application restart [#15](https://github.com/RatakondalaArun/shelf_virtual_directory/issues/15)
  - rootFolderName get failed [#16](https://github.com/RatakondalaArun/shelf_virtual_directory/issues/16)

## 2.0.0

- nullsafety to stable

## 2.0.0-nullsafety.0

- Removed test dependency.
- Migration to nullsafety prerelease [#6](https://github.com/RatakondalaArun/shelf_virtual_directory/issues/6).

## 1.0.1

- bumped shelf version [#4](https://github.com/RatakondalaArun/shelf_virtual_directory/issues/4) by [@YeungKC](https://github.com/YeungKC)

## 1.0.0+1

- updated README.md

## 1.0.0

- Initial Release


test/shelf_virtual_directory_test.dart
import 'dart:io';

import 'package:http/http.dart' as http;
import 'package:path/path.dart' as p;
import 'package:shelf/shelf.dart';
import 'package:shelf/shelf_io.dart' as io;
import 'package:shelf_router/shelf_router.dart';
import 'package:shelf_virtual_directory/shelf_virtual_directory.dart';
import 'package:test/test.dart';

void main() {
  // Tests uses different root directory
  // so we need to append current directory path
  //
  // "web" is the folder that we are trying to serve
  final fsPath = p.join(Directory.current.path, 'example', 'web');

  // final vWebDir = ShelfVirtualDirectory('fsPath');
  test('Throws "ArgumentError" when directory is not found', () {
    expect(() => ShelfVirtualDirectory('doesnotexist'), throwsArgumentError);
  });

  group('Test API calls', () {
    late io.IOServer server;

    setUpAll(() async {
      final virDir = ShelfVirtualDirectory(fsPath);

      final router = Router()
        ..mount('/nodirlisting/',
            ShelfVirtualDirectory(fsPath, listDirectory: false).handler)
        ..mount('/routerstatic/', virDir.router)
        ..mount('/mountstatic/', virDir.handler)
        // only offline tests are possible
        ..mount('/fsrootstatic/', ShelfVirtualDirectory('/').handler)
        // for testing customheader
        ..mount(
            '/customheader/',
            ShelfVirtualDirectory(fsPath, headersParser: _customHeaderParser)
                .handler)
        ..get('/api/user', (_) => Response.ok('/api/user'))
        ..get('/api', (_) => Response.ok('/api'))
        ..mount('/', ShelfVirtualDirectory(fsPath).handler);

      // setup server test files
      final pipline = const Pipeline()
          // .addMiddleware(logRequests())
          .addHandler(router);

      server = await io.IOServer.bind(InternetAddress.loopbackIPv4, 0)
        ..mount(pipline);
    });

    tearDownAll(() => server.close());

    Uri url(String path) => Uri(
          scheme: server.url.scheme,
          userInfo: server.url.userInfo,
          host: server.url.host,
          port: server.url.port,
          path: path,
        );

    // GET Requests
    group('GET Requests', () {
      test('GET ..mount(\'/routerstatic/\') should return 200', () async {
        final res = await http.get(url('/routerstatic/'));
        expect(res.statusCode, equals(200));
      });

      test('GET ..mount(\'/mountstatic/\') should return 200', () async {
        final res = await http.get(url('/mountstatic/'));
        expect(res.statusCode, equals(200));
      });

      test('GET ..mount(\'/fsrootstatic/\') should return 200', () async {
        final res = await http.get(url('/fsrootstatic/'));
        expect(res.statusCode, equals(200));
        // CI services might not have access to root folder so skipping it
      }, skip: true);

      test('GET ..get(\'/api/user\') should return 200', () async {
        final res = await http.get(url('/api/user'));
        expect(res.statusCode, equals(200));
      });

      test('GET Should redirect if there is no trailing slash "/"', () async {
        final req = http.Request('GET', url('/routerstatic/temp'))
          // ..maxRedirects = 1
          ..followRedirects = false;
        final client = http.Client();

        final rs = await client.send(req);
        client.close();
        expect(rs.isRedirect, isTrue);
        expect(rs.statusCode, equals(301));
        expect(rs.headers[HttpHeaders.locationHeader], isNotNull);
        expect(rs.headers[HttpHeaders.locationHeader], endsWith('/'));
      });
    });

    // HEAD requests
    group('HEAD Requests', () {
      test('HEAD ..mount(\'/routerstatic/\') should return 200', () async {
        final res = await http.head(url('/routerstatic/'));
        expect(res.statusCode, equals(200));
      });

      test('HRAD ..mount(\'/mountstatic/\') should return 200', () async {
        final res = await http.head(url('/mountstatic/'));
        expect(res.statusCode, equals(200));
      });

      test('HEAD ..mount(\'/fsrootstatic/\') should return 200', () async {
        final res = await http.head(url('/fsrootstatic/'));
        expect(res.statusCode, equals(200));
        // CI services might not have access to root folder so skipping it
      }, skip: true);

      test('HEAD ..get(\'/api/user\') should return 200', () async {
        final res = await http.head(url('/api/user'));
        expect(res.statusCode, equals(200));
      });

      test('HEAD Should redirect if there is no trailing slash "/"', () async {
        final req = http.Request('HEAD', url('/routerstatic/temp'))
          // ..maxRedirects = 1
          ..followRedirects = false;
        final client = http.Client();

        final rs = await client.send(req);
        client.close();
        expect(rs.isRedirect, isTrue);
        expect(rs.statusCode, equals(301));
        expect(rs.headers[HttpHeaders.locationHeader], isNotNull);
        expect(rs.headers[HttpHeaders.locationHeader], endsWith('/'));
      });
    });
    // Range requests
    group('Range Requests', () {
      // returns partial content
      test('Should response with partial content', () async {
        final res = await http.get(
          url('/temp/BigBuckBunnyTrailer.mp4'),
          headers: {'Range': 'bytes=1933312-2031615'},
        );
        print(res.request!.url.toString());
        expect(res.statusCode, equals(206));
        expect(res.headers[HttpHeaders.contentLengthHeader], equals('98304'));
        expect(res.headers[HttpHeaders.acceptRangesHeader], equals('bytes'));
        expect(res.headers[HttpHeaders.contentRangeHeader], isNotEmpty);
      });
    });
    group('Argument', () {
      test('Should list directory', () async {
        final res = await http.get(url('/routerstatic/temp/'));
        expect(res.statusCode, equals(200));
      });

      test('Should not list directory', () async {
        final res = await http.get(url('/nodirlisting/temp/'));
        expect(res.statusCode, equals(404));
      });

      test('Should contains header in response FileHeaderParser', () async {
        final res = await http.get(url('/customheader/'));
        expect(
          res.headers,
          containsPair('customheader', 'ShelfVirtualDirectory'),
        );
      });
    });
  });
}

Map<String, Object> _customHeaderParser(File file) =>
    {'customheader': 'ShelfVirtualDirectory'};


example/example.dart
import 'dart:io';

import 'package:path/path.dart' as p;
import 'package:shelf/shelf.dart' show Pipeline, Response, logRequests;
import 'package:shelf/shelf_io.dart' as io show serve;
import 'package:shelf_router/shelf_router.dart';
import 'package:shelf_virtual_directory/shelf_virtual_directory.dart';

Future<void> main(List<String> args) async {
  // serving directory
  const folderToServe = 'web';
  final address = InternetAddress.loopbackIPv4;
  final port = int.parse(Platform.environment['PORT'] ?? '8082');

  // creates a [ShelfVirtualDirectory] instance and provides a [Router] instance.
  final folderPath = p.join(
    Directory.current.path,
    'example',
    folderToServe,
  );
  final virDir = ShelfVirtualDirectory(folderPath);

  final apiRouter = Router()
    ..mount('/routerstatic/', virDir.router)
    ..mount('/mountstatic/', virDir.handler)
    ..mount('/nodirlisting/',
        ShelfVirtualDirectory(folderPath, listDirectory: false).handler)
    // ..mount('/fsrootstatic/', ShelfVirtualDirectory('/').handler)
    ..get('/getstatic/', virDir.handler)
    ..get('/api/user', (_) => Response.ok('/api/user'))
    ..get('/api', (_) => Response.ok('/api'))
    ..mount('/', virDir.handler);

  // using [Pipeline] from shelf we can add a logging middleware.
  // we can use handler provided by [Router] instance.
  final pipline =
      const Pipeline().addMiddleware(logRequests()).addHandler(apiRouter);

  // add the handler to [Cascade]
  final server = await io.serve(
    pipline,
    address,
    port,
  );

  print('Server is running at http://${server.address.host}:${server.port}');
}


lib/src/html_templates.dart
part of shelf_virtual_directory;

String _listDirhtmlStart(String heading, String requestedPath) => '''
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>$heading/</title>
    <style>
      h1 {
        width: 100%;
        font-size: 20px;
        background-color: #4361ee;
        color: #edf2fb;
        padding-top: 10px;
        padding-bottom: 10px;
        padding-left: 5px;
        padding-right: 5px;
      }
      li {
        font-size: 18px;
      }
      ul {
        display: block;
        list-style-type: disc;
        margin-block-start: 5px;
        margin-block-end: 5px;
        margin-inline-start: 0px;
        margin-inline-end: 0px;
        padding-inline-start: 10px;
      }
      .container {
        display: flex;
        flex-direction: column;
        justify-content: center;
      }
      /*mobile*/
      @media screen and (max-width: 450px) {
        .container {
          width: 90%;
          padding-left: 5%;
          padding-right: 5%;
        }
      }
      /*tablet*/
      @media screen and (min-width: 451px) and (max-width: 700px) {
        .container {
          width: 90%;
          padding-left: 5%;
          padding-right: 5%;
        }
      }
      /*PC*/
      @media screen and (min-width: 701px) and (max-width: 900px) {
        .container {
          width: 80%;
          padding-left: 10%;
          padding-right: 10%;
        }
      }
      @media screen and (min-width: 900px) {
        .container {
          width: 80%;
          padding-left: 10%;
          padding-right: 10%;
        }
      }
      table {
        border-collapse: collapse;
      }
      td,
      th {
        border: none;
        text-align: left;
        padding: 8px;
      }

      tr:nth-child(even) {
        background-color: #dddddd;
      }
    </style>
  </head>
  <body>
    <div class="container">
      <h1>📁$requestedPath</h1>
      <table>
        <tr>
          <th>Name</th>
          <th>Modified</th>
          <th>Permission</th>
          <th>Size</th>
        </tr>
''';

String _tableEnd() => '</table>';

String _listDirHtmlEnd() => '''
      <div>
        <a href="https://pub.dev/packages/shelf_virtual_directory">shelf_virtual_directory</a>
      </div>
    </div>
  </body>
</html>

''';


lib/src/virtual_directory.dart
library shelf_virtual_directory;

import 'dart:async';
import 'dart:convert';
import 'dart:io';

import 'package:http_parser/http_parser.dart' as hp;
import 'package:mime/mime.dart' as mime;
import 'package:path/path.dart' as path;
import 'package:shelf/shelf.dart';
import 'package:shelf_router/shelf_router.dart';

part 'html_templates.dart';

/// Creates a instance of [ShelfVirtualDirectory]
///
/// ## Parameters
/// - `folderPath`: Name of the directory you want to serve from the *current folder*
///
/// - `defaultFile`: File name that will be served. *Default: index.html*
///
/// - `default404File`: File name that will be served for 404. *Default: 404.html*
///
/// - `showLogs`: Shows logs from the ShelfVirtualDirectory. *Default: true*
///
/// ## Examples
///
/// You can get router or handler or cascade from [ShelfVirtualDirectory]
/// instance
/// ```dart
///
/// final virDirRouter = ShelfVirtualDirectory(folderToServe);
///
/// final staticFileHandler = const Pipeline()
///     .addMiddleware(logRequests())
///     .addHandler(virDirRouter.handler);
///
/// io.serve(Cascade().add(staticFileHandler).handler,address,port).then((server){
///     print('Server is sunning at ${server.address}:${server.port}'),
/// });
///```
class ShelfVirtualDirectory {
  /// Absolute path to the directory
  final String folderPath;

  /// Default file to serve. Must be inside [folderPath].
  final String defaultFile;

  /// Default 404 file to serve. Must be inside [folderPath].
  final String default404File;

  /// Weather to list directory or not.
  final bool listDirectory;

  final FileHeaderParser _headersParser;
  final Directory _dir;

  /// Creates a instance of [ShelfVirtualDirectory]
  ///
  /// ## Parameters
  ///
  /// - `folderPath`: Name of the directory you want to serve. *Must be a absolute path*.
  /// - `defaultFile`: File name that will be serve. *Must be inside [folderPath]*
  /// - `default404File`: File name that will be served for 404. *Default: 404.html*
  /// - `listDirectory`: Lists files and directories from the [folderPath]
  /// - `headersParser`: Provide your own headers from the [File].
  ///
  /// ## Examples
  ///
  /// You can get router or handler or cascade from [ShelfVirtualDirectory]
  /// instance
  /// ```dart
  ///
  /// final virDirRouter = ShelfVirtualDirectory(folderToServe);
  ///
  /// final staticFileHandler = const Pipeline()
  ///     .addMiddleware(logRequests())
  ///     .addHandler(virDirRouter.handler);
  ///
  /// io.serve(Cascade().add(staticFileHandler).handler,address,port).then((server){
  ///     print('Server is sunning at ${server.address}:${server.port}'),
  /// });
  ///```
  ShelfVirtualDirectory(
    this.folderPath, {
    this.defaultFile = 'index.html',
    this.default404File = '404.html',
    this.listDirectory = true,
    FileHeaderParser headersParser = _defaultFileheaderParser,
  })  : _dir = Directory(folderPath),
        _headersParser = headersParser {
    if (!_dir.existsSync()) {
      throw ArgumentError('A directory corresponding to folderpath '
          '"$folderPath" could not be found');
    }
  }

  /// Creates a instance of [Handler]
  Handler get handler => _handler;

  /// Returns a instance of [Router]
  ///
  /// Can be used to mount as a subroute
  ///
  /// ```
  /// final router = Router('/',ShelfVirtualDirectory('web').router);// localhost:8080/
  /// //or
  /// final router = Router('/home/',ShelfVirtualDirectory('web').router);//localhost:8080/home/
  /// ```
  Router get router => Router(notFoundHandler: _handler);

  /// Returns a instance of [Cascade]
  ///
  /// Can be used to directly serve from server
  /// ```
  /// import 'package:shelf/shelf_io.dart' as io show serve;
  ///
  /// final cascade = ShelfVirtualDirectory('web').cascade;
  /// io.serve(cascade,'localhost',8080).then((server){
  ///   print('Server is sunning at ${server.address}:${server.port}'),
  /// })
  ///
  /// ```
  Cascade get cascade => Cascade().add(_handler);

  Future<Response> _handler(Request req) async {
    if (!['GET', 'HEAD'].contains(req.method)) {
      return Response.notFound('Not Found');
    }

    final basePath = await _dir.resolveSymbolicLinks();

    // file system path for this entity
    final fsPath = path.joinAll([basePath, ...req.url.pathSegments]);

    // checks if index file should be served
    final fsPathType = await FileSystemEntity.type(fsPath);

    // if (fsPathType == FileSystemEntityType.notFound) {
    //   final indexFile = File(path.join(basePath, defaultFile));
    //   if (!await indexFile.exists()) return Response.notFound('Not Found');
    //   final headers = await _headersParser(indexFile);
    //   return Response(200, body: indexFile.openRead(), headers: headers);
    // }

    // checks if we should serve index file aka `defaultFile`
    if (req.url.path.isEmpty) {
      final indexFile = File(path.join(basePath, defaultFile));
      // if index file exists in that directory serve it
      if (await indexFile.exists()) {
        final headers = await _headersParser(indexFile);
        return Response(
          200,
          body: req.method == 'HEAD' ? null : indexFile.openRead(),
          headers: headers,
        );
      }
      return _handleDir(
        req,
        basePath,
        Directory(fsPath),
        _headersParser,
      );
    }

    switch (fsPathType) {
      case FileSystemEntityType.link:
        // todo: handle symboliclinks
        stdout
          ..write('[ShelfVirtualDirectory] ')
          ..write('Resolving symbolic link was not implemented ')
          ..write('Raise a issue at ')
          ..write('https://github.com/RatakondalaArun/')
          ..write('shelf_virtual_directory/issues/new')
          ..write('?title=Resolving%20Symbolic%20link%20failed\n');
        return _handleFile(req, basePath, null, _headersParser);
      case FileSystemEntityType.directory:
        return _handleDir(
          req,
          basePath,
          Directory(fsPath),
          _headersParser,
        );
      case FileSystemEntityType.file:
        return _handleFile(req, basePath, File(fsPath), _headersParser);
      default:
        return _handleFile(req, basePath, null, _headersParser);
    }
  }

  Future<Response> _handleFile(
    Request req,
    String fsPath,
    File? file,
    FileHeaderParser headerParser,
  ) async {
    // serves default404file incase requested file does not exist
    if (file == null) {
      final notFoundFile = File(path.join(fsPath, default404File));
      if (!await notFoundFile.exists()) return Response.notFound('Not Found');

      return Response.notFound(
        notFoundFile.openRead(),
        headers: await _headersParser(notFoundFile),
      );
    }

    // collect file data
    final fileStat = await file.stat();

    // check file permission of
    if (fileStat.modeString()[0] != 'r') return Response.forbidden('Forbidden');
    final length = fileStat.size;
    final range = req.headers[HttpHeaders.rangeHeader];
    final headers = await headerParser(file) ?? {};

    if (range != null) {
      final matches = RegExp(r'^bytes=(\d*)\-(\d*)$').firstMatch(range);
      if (matches != null) {
        final startMatch = matches[1] ?? '';
        final endMatch = matches[2] ?? '';
        if (startMatch.isNotEmpty || endMatch.isNotEmpty) {
          int start;
          int end;
          if (startMatch.isEmpty) {
            start = length - int.parse(endMatch);
            if (start < 0) start = 0;
            end = length - 1;
          } else {
            start = int.parse(startMatch);
            end = endMatch.isEmpty ? length - 1 : int.parse(endMatch);
          }
          // If the range is syntactically invalid the Range header
          // MUST be ignored (RFC 2616 section 14.35.1).
          if (start <= end) {
            if (end >= length) {
              end = length - 1;
            }
            if (start >= length) {
              return Response(
                HttpStatus.requestedRangeNotSatisfiable,
                // headers: headers,
              );
            }

            // Override Content-Length with the actual bytes sent.
            headers[HttpHeaders.contentLengthHeader] =
                (end - start + 1).toString();

            // Set 'Partial Content' status code.
            headers[HttpHeaders.contentRangeHeader] =
                'bytes $start-$end/$length';
            // Pipe the 'range' of the file.

            return Response(
              HttpStatus.partialContent,
              body: req.method == 'HEAD' ? null : file.openRead(start, end + 1),
              headers: headers,
            );
          }
        }
      }
    }

    return Response(
      200,
      body: req.method == 'HEAD' ? null : file.openRead(),
      headers: headers,
    );
  }

  Future<Response> _handleDir(
    Request req,
    String fsPath,
    Directory dir,
    FileHeaderParser headerParser,
  ) async {
    final requestedPath = req.url.path;
    if (!listDirectory) return Response.notFound('Not Found');
    if (!requestedPath.endsWith('/') && requestedPath.isNotEmpty) {
      return Response.movedPermanently('${req.requestedUri.toString()}/');
    }

    final controller = StreamController<List<int>>();
    const encoding = Utf8Codec();
    const sanitizer = HtmlEscape();

    void add(String string) {
      controller.add(encoding.encode(string));
    }

    add(_listDirhtmlStart(
      sanitizer.convert(Uri.file(dir.path).pathSegments.last),
      sanitizer.convert(requestedPath),
    ));

    final segments = requestedPath.split('/');
    if (segments.isNotEmpty && segments.length > 2) {
      add('''
  <tr>
    <td><a style="cursor: pointer;" onclick="window.location.href = '../'">..</a></td>
    <td>-</td>
    <td>-</td>
    <td>-</td>
  </tr>
''');
    }

    try {
      final subEntities = await dir.list().toList();
      // sort entities
      subEntities.sort((e1, e2) {
        if (e1 is Directory && e2 is! Directory) {
          return -1;
        }
        if (e1 is! Directory && e2 is Directory) {
          return 1;
        }
        return e1.path.compareTo(e2.path);
      });

      for (var subEntity in subEntities) {
        final entityStat = await subEntity.stat();
        final entityName = Uri.file(subEntity.path).pathSegments.last;
        add(_tr(entityName, requestedPath, entityStat));
      }

      add(_tableEnd());
      add('<p>total: ${subEntities.length}</p>');
      add(_listDirHtmlEnd());
      // ignore: unawaited
      controller.close();

      return Response.ok(
        controller.stream,
        encoding: encoding,
        headers: {HttpHeaders.contentTypeHeader: 'text/html'},
      );
    } on FileSystemException catch (err) {
      stderr
        ..writeln(err)
        ..writeln(StackTrace.current);
      return Response.forbidden('Forbidden');
    } catch (e) {
      rethrow;
    }
  }
}

// Parse header and return headers for the file
typedef FileHeaderParser = FutureOr<Map<String, Object>?> Function(File file);

Future<Map<String, Object>> _defaultFileheaderParser(File file) async {
  final fileType = mime.lookupMimeType(file.path);

  // collect file data
  final fileStat = await file.stat();

  // check file permission
  if (fileStat.modeString()[0] != 'r') return {};

  return {
    HttpHeaders.contentTypeHeader: fileType ?? 'application/octet-stream',
    HttpHeaders.contentLengthHeader: fileStat.size.toString(),
    HttpHeaders.lastModifiedHeader: hp.formatHttpDate(fileStat.modified),
    HttpHeaders.acceptRangesHeader: 'bytes'
  };
}

/// Table row
String _tr(String name, String requestedPath, FileStat stat) {
  final isDir = stat.type == FileSystemEntityType.directory;
  final modified = stat.modified.toLocal();
  return '''
  <tr>
    <td><small>${isDir ? '📁' : '📄'}</small><a href="$name${isDir ? '/' : ''}">$name ${isDir ? '/' : ''}</a></td>
    <td>${modified.day}/${modified.month}/${modified.year} ${modified.hour}:${modified.minute}</td>
    <td>${stat.modeString()}</td>
    <td>${stat.size / 100} kb</td>
  </tr>
''';
}
